website/source/docs/guides/consul-splitting.md

```diff
@@ -0,0 +1,461 @@
+
+1. Set up a traffic split to make sure v2 doesn’t receive any traffic at first.
+1. Start an instance of the v2 API service in your production environment.
+1. Register v2 so that Consul can send traffic to it.
+1. Slowly shift traffic to v2 and a way from v1 until the new version is
+handling all the traffic.
+
+## Configure Traffic Splitting
+
+Traffic Splitting uses configuration entries (introduced in Consul 1.5 and 1.6)
+to centrally configure the services and Envoy proxies. There are three
+configuration entries you need to create to enable traffic splitting:
+
+- Service Defaults for the API service to set the protocol to HTTP.
+- Service Splitter which defines the traffic split between the service subsets.
+- Service Resolver which defines which service instances are version 1 and 2.
+
+### Configuring Service Defaults
+
+Traffic splitting requires that the upstream application uses HTTP, because
+splitting happens on layer 7 (on a request by request basis). You will tell
+Consul that your upstream service uses HTTP by setting the protocol in a
+“service defaults” configuration entry for the API service. This configuration
+is already in your demo environment at `l7_config/api_service_defaults.json`. It
+contains the following.
+
+```json
+{
+  "kind": "service-defaults",
+  "name": "api",
+  "protocol": "http"
+}
+```
+
+The `kind` field denotes the type of configuration entry which you are defining;
+for this example, `service-defaults`. The `name` field defines which service the
+service-defaults configuration entry applies to. (The value of this field must
+match the name of a service registered in Consul, in this example, `api`.) The
+`protocol` is `http`.
+
+To apply the configuration, you can either use the Consul CLI or the API. In
+this example we’ll use the configuration entry endpoint of the HTTP API, which
+is available at `http://localhost:8500/v1/config`. To apply the config, use a
+PUT operation in the following  command.
+
+```shell
+$ curl localhost:8500/v1/config -XPUT -d @l7_config/api_service_defaults.json
+true%
+```
+
+Find more information on service-defaults configuration entries in the
+[documentation](https://www.consul.io/docs/agent/config-entries/service-defaults.html).
+
+### Configuring the Service Resolver
+
+The next configuration entry you need to add is the Service Resolver, which
+allows you to define how Consul’s service discovery selects service instances
+for a given service name.
+
+Service Resolvers allow you to filter for subsets of services based on
+information in the service registration. In this example, we are going to define
+the subsets “v1” and “v2” for the API service, based on its registered metadata.
+API service version 1 in the demo is already registered with the tags `v1` and
+service metadata `version:1`. When you register version 2 you will give it the
+tag `v2` and the metadata `version:2`. The `name` field is set to the name of
```

Comment:
Hmm is this because the UI only shows tags right now and later you use that to show the difference?

I guess that's reasonable but maybe we should be super transparent about that so it's clear that tags are not actually needed for this splitting config to work?

@opihana @johncowen not sure if this is on you're radar although I feel like it's been mentioned in passing before but we should consider how the UI can use service meta in a similar way to how it surfaces tags currently to make this not a reason to still use tags going forward?

Reply:
FWIW the behavior that I originally expected was that you could filter directly on tags. I haven't tried this, so maybe it's possible and I just don't know?

Reply:
Yeah you definitely can filter just on tags it would look like `v1 in Service.Tags` in the subset filter string (not tested but pretty sure that's right).

I think the question is whether we want to promote tag usage as we'd kinda like to deprecate it but currently the UI only shows tags... I'm OK only using tags now if it makes this guide simpler but would be nice to update to use meta only once we have a UI solution that allows meta-only service instances to have some meta visible easily in the UI?

Reply:
Hey folks 👋 

TLDR; (it got long 😆 ) In my mind tags are for humans, I think we should recommend using meta data for splitting. Metadata is surfaced in the UI but we need to make it easier to get to.

---

Yeah tags are a funny one, and they are becoming more 'funny peculiar' as we move into the discovery chain/resolving work.

So a bit of background: Specifically in the UI we 'rollup' tags on your Services on the service listing page, rollup referring to the fact that we aggregate all the tags from all the Service Instances of a particular Service into the same list. That means that if you have some Service Instances with the tag `v1` and some with the tag `v2`, when you look at the Service listing the Service has _both_ `v1` and `v2`.

This isn't a UI thing exactly as it comes from the API like this, i.e. we don't aggregate this on the frontend. For me this is sometimes confusing, yet kind of 'correct'.

We've spoken quite a lot about using/not using 'rollups' (aggregating Service Instance data into the Service) for surfacing things like this. Ideally I'd fall down on the side of moving away from rolling-up unless we can make it so the user can explicitly say 'this _entire_ Service has the tag v1', this maybe achievable via centralized config in a very distant future. Also, 'ideally' isn't something we can always do as it breaks workflows for others.

---

My 2 cents on what we could do: As @banks mentioned we are kind of 'deprecating' tags, tags being just a type of metadata that helps to describe an item and makes it easy to find again at a later date. I personally feel that we should use Meta itself for anything 'technical/machine' related moving forwards which I think is what @banks is suggesting, and _if anything_ use Tags for 'humans' to enable them to find things (getting very specific this is what a tag is for, making it easier for humans to find things) either via the UI or something else. So for this exact case ^ in the docs, maybe just use Meta in the examples but mention below as an aside that you can filter on other things such as Tags (and whatever else folks can filter on)? 

From a UI standpoint, we recently added Meta to the UI, but you can only see this if click into the Service Instance and then into a Meta Data tab (i.e. we don't 'roll them up' currently like we do with tags), this is still 'correct' but less confusing when working with the discovery chain - and is four clicks away from the Service listing! 🙀 

I would imagine people immediately reach for tags to further categorise Services partly because they are more visible in the UI, so I definitely think we should make Meta much easier to reach and browse in the UI and I think there are a few options that we've spoken about before and I think we should see if we can pick these discussions back up. @opihana , @jnwright I'm gonna add this to our agenda for our next catch up but obvs feel free to pitch in here also.